File: jupytershim/notebook_comm.py
from uuid import uuid4
from websocket import WebSocket
import json
import time

WSJS_TMPL = """
<script>
    var execution_id = "%s";                                                 // Zeppelin double print workaround
    if(window.__zeppelin_already_executed__ == null) {                       // Zeppelin double print workaround
        window.__zeppelin_already_executed__ = [];                           // Zeppelin double print workaround
    }                                                                        // Zeppelin double print workaround
    if(!window.__zeppelin_already_executed__.includes(execution_id)) {       // Zeppelin double print workaround

        var $scope = angular.element(document.getElementById("%s")).scope();

        if(typeof(window.__jupyter_notebook_unwatchers__) !== "undefined") {
            console.info("NoteboooComm: cancel watchers");
            var unwatchers = window.__jupyter_notebook_unwatchers__
            for(i in unwatchers) {
                unwatchers[i]();
            }
        }

        if((typeof(Jupyter) === "undefined") || (Jupyter === null)) {
            console.info("Initiate Javascript Notebook Comms");
            Jupyter = {};
            Jupyter.notebook = new Notebook(Jupyter);
        }

        window.__jupyter_notebook_unwatchers__ = [];
        
        var unwatch = $scope.$watch("__jupyter_comm_msg__", function(newValue, oldValue, scope) {
            if(typeof(newValue) !== "undefined") {
                // console.info("__jupyter_comm_msg__: " + JSON.stringify(newValue));
                Jupyter.notebook.kernel.notebookComm.handleMsg(newValue);
            }
        }, true)
        window.__jupyter_notebook_unwatchers__.push(unwatch)

        window.__zeppelin_already_executed__.push(execution_id);             // Zeppelin double print workaround
    } else {
        console.info("zeppelin bug, angular script already executed, skipped");
    }
</script>
"""

class ZeppelinNotebookComm:

    def __init__(self, jupyterShim, zeppelinContext, debug=False):
        self.id = 0
        self.z = zeppelinContext
        self.jupyterShim = jupyterShim
        self.notebookCommDivId = "__Jupyter_Notebook_Comm__"
        self.reset()
        
        # div must exist before javascript below can be printed
        print("""%%angular <div id="%s">Shim initialized (do not delete this paragraph)</div>""" % self.notebookCommDivId)
        
        if debug:
            print("""%angular Debug: {{__jupyter_comm_msg__}}""")
            
        print("%angular") 
        print(WSJS_TMPL % (str(uuid4()), self.notebookCommDivId))
        
    def send(self, task, msg):
        self.id += 1
        self.z.angularBind("__jupyter_comm_msg__", {"task":task, "msg":msg})
        
    def reset(self):
        self.z.angularUnbind("__jupyter_comm_msg__")
        self.id = 0

File: jupytershim/js/notebook_comm.js
/**
  *  class ZeppelinNotebookComm
  */

var ZeppelinNotebookComm = function(kernel) {
    console.info("ZeppelinNotebookComm: init");
    this.kernel = kernel;
}   

ZeppelinNotebookComm.prototype.handleMsg = function(data) {
    if(data.task == "publish") {
        this.publish(data.msg.div_id, data.msg.html);
    } else if(data.task == "comm_open") {
        var msg = data.msg;
        this.kernel.comm_manager.new_comm(msg.target_name, msg.data, {}, msg.metadata, msg.comm_id)
    } else if(data.task == "comm_close") {
        console.log("comm_close:");
        console.log(data);
    } else if(data.task == "comm_msg") {
        var msg = data.msg;
        this.kernel.comm_manager.comms[msg.comm_id].then(function(comm) {
            comm.handle_msg({"content": msg}); 
        });
    } else {
        console.error("UNHANDLED:" + JSON.stringify(data));
    }
}

ZeppelinNotebookComm.prototype.send = function(task, msg) {
    console.info("ZeppelinNotebookComm: send " + task);
    this.zeppelinWS.send(JSON.stringify({"notebook_id":this.notebookId, "task": task, "node":"notebook", "msg":msg}));
}

ZeppelinNotebookComm.prototype.publish = function(div_id, html) {
    console.info("ZeppelinNotebookComm: publish " + div_id);
    setTimeout(function(){
        div = document.getElementById(div_id);
        div.innerHTML = html;
    }, 100);
}